#!/usr/bin/env bash

#########################################################################
##                                                                     ##
##     aligned_fasta.sh                                                ##
##                                                                     ##
##     Gael A. Millot                                                  ##
##     Bioinformatics and Biostatistics Hub                            ##
##     Institut Pasteur Paris                                          ##
##                                                                     ##
#########################################################################




select_ch=${1}
igblast_aa=${2}

if (( $(cat ${select_ch} | wc -l ) <= 1 )) ; then
    echo -e "\n\n========\n\nERROR IN NEXTFLOW EXECUTION OF THE aligned_fasta PROCESS\n\nEMPTY INPUT FILE (GENERATED BY THE parseDb_filtering PROCESS)\n\n========\n\n"
    exit 1
fi

# make fasta files of the aligned sequences
mkdir aligned_seq
if [[ "${igblast_aa}" == "false" ]] ; then
    TEMPO_NAME="sequence_alignment"
else 
    TEMPO_NAME="sequence_aa_alignment"
fi
awk -v var1=$igblast_aa -v var2=${TEMPO_NAME} 'BEGIN{FS="\t" ; ORS="\n" ; OFS="\t"}{
    if(NR==1){
        COL_NAME="FALSE"
        COL_SEQ="FALSE"
        for(i4=1; i4<=NF; i4++){
            if($i4=="sequence_id"){COL_NAME=i4}
            if($i4==var2){COL_SEQ=i4}
        }
        if(COL_NAME=="FALSE"){
            print "\n\n========\n\nERROR IN NEXTFLOW EXECUTION OF THE aligned_fasta PROCESS\n\nNO sequence_id COLUMN NAME FOUND IN THE INPUT FILE\n\n========\n\n"
            exit 1
        }
        if(COL_SEQ=="FALSE"){
            print "\n\n========\n\nERROR IN NEXTFLOW EXECUTION OF THE aligned_fasta PROCESS\n\nNO "var2" COLUMN NAME FOUND IN THE INPUT FILE\n\n========\n\n"
            exit 1
        }
    }else{
        if($COL_SEQ!~/^[-NATGC.]*$/ && var1=="false"){
            print "\n\n========\n\nERROR IN NEXTFLOW EXECUTION OF THE aligned_fasta PROCESS\n\n"var2" COLUMN NAME OF THE INPUT FILE MUST BE A NUCLEOTIDE SEQUENCE IF THE igblast_aa PARAMETER IS SET TO false\nHERE IT SEEMS TO BE MADE OF AMINO ACIDS:\n"$COL_SEQ"\n\n========\n\n"
            exit 1
        }
        if($COL_SEQ~/^[-NATGC.]*$/ && var1=="true"){
            print "\n\n========\n\nERROR IN NEXTFLOW EXECUTION OF THE aligned_fasta PROCESS\n\n"var2" COLUMN NAME OF THE INPUT FILE MUST BE A AMINO ACID SEQUENCE IF THE igblast_aa PARAMETER IS SET TO true\nHERE IT SEEMS TO BE MADE OF NUCLEOTIDS:\n"$COL_SEQ"\n\n========\n\n"
            exit 1
        }
        print ">"$COL_NAME"\n"$COL_SEQ > "./aligned_seq/"$COL_NAME".fasta" # make fasta files of the aligned nuc sequences
    }
}' ${select_ch} |& tee -a aligned_fasta.log
# end make fasta files of the aligned sequences





